src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java

```diff
@@ -0,0 +1,218 @@
+///////////////////////////////////////////////////////////////////////////////////////////////
+
+package com.puppycrawl.tools.checkstyle.checks.javadoc;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.puppycrawl.tools.checkstyle.GlobalStatefulCheck;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.DetailNode;
+import com.puppycrawl.tools.checkstyle.api.JavadocTokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.CommonUtil;
+
+/**
+ * <p>
+ * Checks the alignment of
+ * <a href="https://docs.oracle.com/en/java/javase/14/docs/specs/javadoc/doc-comment-spec.html#leading-asterisks">
+ * leading asterisks</a> in a Javadoc comment. The Check ensures that leading asterisks
+ * are aligned vertically under the second asterisk ( &#42; )
+ * of opening Javadoc tag. The alignment of closing Javadoc tag ( &#42;/ ) is also checked.
+ * If a line contains a non-whitespace character before the leading asterisk then that line will be
+ * ignored. Same applies for closing Javadoc tag.
+ * If the ending javadoc line contains a leading asterisk, then that leading asterisk's alignment
+ * will be considered, the closing Javadoc tag will be ignored.
+ * </p>
+ * <p>
+ * If you're using tabs then specify the the tab width in the
+ * <a href="https://checkstyle.org/config.html#tabWidth">tabWidth</a> property.
+ * </p>
+ * <ul>
+ * <li>
+ * Property {@code violateExecutionOnNonTightHtml} - Control when to print violations if the
+ * Javadoc being examined by this check violates the tight html rules defined at
+ * <a href="https://checkstyle.org/writingjavadocchecks.html#Tight-HTML_rules">Tight-HTML Rules</a>.
+ * Type is {@code boolean}.
+ * Default value is {@code false}.
+ * </li>
+ * </ul>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code javadoc.asterisk.indentation}
+ * </li>
+ * <li>
+ * {@code javadoc.missed.html.close}
+ * </li>
+ * <li>
+ * {@code javadoc.parse.rule.error}
+ * </li>
+ * <li>
+ * {@code javadoc.unclosedHtml}
+ * </li>
+ * <li>
+ * {@code javadoc.wrong.singleton.html.tag}
+ * </li>
+ * </ul>
+ *
+ * @since 10.17.0
+ */
+@GlobalStatefulCheck
+public class JavadocLeadingAsteriskAlignCheck extends AbstractJavadocCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "javadoc.asterisk.indentation";
+
+    /** Specifies the column number of starting block of the javadoc comment with tabs expanded. */
+    private int commentBeginColNumber;
+
+    /** Specifies the line number of starting block of the javadoc comment. */
+    private int commentBeginLineNumber;
+
+    /**
+     * Specifies the column number of the leading asterisk
+     * where tabs are converted to a single whitespace.
+     */
+    private int singleSpaceAsteriskColNumber;
```

Comment:
We're firstly checking the alignment of the leading asterisk with expanded tabs. We've used following method for expanding the tabs:

https://github.com/checkstyle/checkstyle/blob/d670bc3bc382150dd022907fec3239f693e296b4/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L162


We need to check the alignment with expanded tabs otherwise the check will not be able to log violations correctly. 

But the `log()` method expects column number without expanded tabs, it internally expands tabs so we need to provide the column number without expanding tabs, `singleSpaceAsteriskColNumber` is the field which stores the column number without expanded tabs. Then we're providing it to the log method:

https://github.com/checkstyle/checkstyle/blob/d670bc3bc382150dd022907fec3239f693e296b4/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L209-L214

The other field which also stores column number is:

https://github.com/checkstyle/checkstyle/blob/d670bc3bc382150dd022907fec3239f693e296b4/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L91-L92

This field stores the column number of the 2nd `*` of starting javadoc block ( /** ) with expanded tabs, then it is provided to the log method to print on which column number the leading asterisk should be placed.

Reply:
Ok, then let's call this one `expectedColumnNumber`.

Reply:
That name is already occupied here:

https://github.com/checkstyle/checkstyle/blob/3be8df24f9d69d137850127a30ca8f56504c44b8/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L93

Can we call is `asteriskColumnNumberWithoutTabs` ?

Reply:
NVM, renamed `singleSpaceAsteriskColNumber` to: https://github.com/checkstyle/checkstyle/blob/77dd1d104385ad79133495efee9f504f23f06730/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocLeadingAsteriskAlignCheck.java#L103